### ISSUE TITLE
Support truck (HGV) routing properties

### ISSUE THREAD

**Issue by nvkelso** [tilezen/vector-datasource#1553]
**tl;dr**

See [PR #1709](https://github.com/tilezen/vector-datasource/pull/1709) for documentation on what was actually decided. 

I'm fine if this is for zooms 14+ only at first? We should generally follow the stripping we added in v1.6.0 depending on the kind of road...

- [ ] Add `low_emission_zone` polygons to **landuse** layer.
- Add properties on existing line geoms in the **roads** layer:
    - [ ] `hgv`: optional property indicating general truck heavy goods vehicle truck access. See below for list of values.
        values whitelist: `no`, `designated`, `destination`, `delivery`, `local` and `agricultural`.
    - [ ] `hgv_restriction`: optional property indicating limitations to heavy goods vehicle truck access. See below for list of values. Available on both point and line geometries. See also `hgv_restriction_shield_text`.
        values include: `weight` (kg), `height` (cm), `length` (cm), `width` (cm), `wpa` (weight per axle, in kg), `kpra` (king pin to rear axle leght, in cm), `hazmat`, `other` and `multiple` if more than one.
    - [ ] `hgv_restriction_shield_text`: optional and paired with `hgv_restriction` points with values like `5.1m`. Because the units are different per restriction an abbreviation should be provided. Values in meters can be specified with one decimal precision but value of 5.0m should be given as 5m.
    - [ ] `toll`: optional `boolean` value indicating whether a fee must be paid by general traffic to travel the feature.
    - [ ] `toll_hgv`: optional `boolean` value when tool only applies to certain classes of vehicles like heavy good vehicle trucks.
- Add new `hgv_restriction` kind points in the **roads** layer, with following properties:
    - [ ] `hgv`: optional property indicating general truck heavy goods vehicle truck access. 
        values whitelist: `no`, `designated`, `destination`, `delivery`, `local` and `agricultural`.
    - [ ] `hgv_restriction`: optional property indicating limitations to heavy goods vehicle truck access. See below for list of values. Available on both point and line geometries. See also `hgv_restriction_shield_text`.
        values include: `weight` (kg), `height` (cm), `length` (cm), `width` (cm), `wpa` (weight per axle, in kg), `kpra` (king pin to rear axle leght, in cm), `hazmat`, `other` and `multiple` if more than one.
    - [ ] `hgv_restriction_shield_text`: optional and paired with `hgv_restriction` points with values like `5.1m`. Because the units are different per restriction an abbreviation should be provided. Values in meters can be specified with one decimal precision but value of 5.0m should be given as 5m.
    - [ ] Possibly update documentation to clarify that `hgv_restriction` is an allowed kind in the roads layer?

===

Just like we did for Bike Map support, let's add properties on roads layer to parallel what Valhalla sees from OSM for `truck_speed` and `truck_route`. There may also be additional properties for height restrictions on bridges and such?

See also:

- https://github.com/valhalla/valhalla-docs/blob/master/map-matching/api-reference.md
- https://github.com/valhalla/valhalla/blob/537146b6d978b6e70de12f0542909880fac871ae/src/sif/truckcost.cc
- https://wiki.openstreetmap.org/wiki/Key:hgv
- https://taginfo.openstreetmap.org/keys/hgv#map

**Comment by musculman** [comment]
I believe that these properties should not be called truck routing but rather just truck attributes or truck properties, as they are used for routing but also for visualization.
Every link of a particular road could contain one or more of the so-called truck attributes.
Please consider if these list of attributes would make sense:

1. truck restriction
2. weight restriction in kg
3. height restriction in cm
4. length restriction in cm
5. width restriction in cm
6. WPA (Weight per Axle in kg) restriction
7. KPRA (King pin to Rear Axle Lenght in cm) restriction
8. Preferred Route
9. Hazardous Materials Permit Required (usually refer to the type of permit that is required to legally drive on this road with hazardous cargo)
10. Truck Speed Limit in km/h
11. Transport Verified
12. Environmental Zone
13. Truck Toll (for example the [ÉcoTaxe of France](https://fr.wikipedia.org/wiki/Taxe_nationale_sur_les_v%C3%A9hicules_de_transport_de_marchandise))

Please note most of these restrictions should support multiple values and they are not only boolean or single valued ones.
Some of them are simple, like:
```
width restriction
general trailer: 2.55m
refrigerated vehicle (minimum thickness of walls - 45mm): 2.60m
```

A bit of research is required for some attributes and what capabilities should they have.
For example, the WPA restriction can have values that need multiple options such as:

```
weight per axle on the main road network
single axle 10.0t
drive axle 11.5t
tandem drive axle with a distance between the axles of
- 1.20m - 1.29m 16.0t
- 1.30m - 1.80m 18.0t
- >1.80m 20.0t
```

**Comment by musculman** [comment]
As there has not been feedback on this topic, I will try to draft a first version of it, to try to put the concepts more clear. Perhaps it was just too confusing.

**Comment by musculman** [comment]
Question that came up while drafting the content: should this data come in a separated new layer so that we can still merge and optimize the road segments or should we just add more attributes on the road layer?
One could argue in favour of the first option because otherwise you cannot merge all road segments when simplifying geometries, on the other hand, you are duplicating geometries and there is more work on the client side if you provide two layers (roads and truck attributes) with road network.

Thoughts?

**Comment by nvkelso** [comment]
Hi @musculman. We already support some visualization of bicycle routing and bus properties on the roads... but only the ones that make sense for **visual display**. 

For example: `Truck Speed Limit in km/h` is just used for **routing** and is probably not appropriate for the vector tiles.

I can see including these properties:

- **truck_restriction** `boolean`
    - if any weight, height, length, width & etc?
- **truck_network**
    - boolean, or would this be more than one route?
    - we'd generally say `truck_network` and `all_truck_networks` for general route networks, bicycle networks, and bus networks.
- **truck_hazmat_permit**
    - Hazardous Materials Permit Required boolean?
    - (usually refer to the type of permit that is required to legally drive on this road with hazardous cargo)

Maybe these as optional point geometry in roads layer where the truck restriction starts/stops?

- `truck_restriction` kind with _kind_detail_ of **weight**
- `truck_restriction` kind with _kind_detail_ of **height**
- `truck_restriction` kind with _kind_detail_ of **length**
- `truck_restriction` kind with _kind_detail_ of **width**
- `truck_restriction` kind with _kind_detail_ of **wpa**
- `truck_restriction` kind with _kind_detail_ of **kpra**

Where kind_detail is the dominant restriction used for symbolizing the point... and if there is more than one restriction then properties can clarify?

- `truck_restriction_weight` (in kg)
- `truck_restriction_height` (in cm)
- `truck_restriction_length` (in cm)
- `truck_restriction_width` (in cm)
- `truck_restriction_wpa` (Weight per Axle in kg)
- `truck_restriction_kpra` (King pin to Rear Axle Lenght in cm)

This seems in-appropriate as it's routing only:

- Truck Speed Limit in km/h

Don't know enough about these, but they also seem routing only:

- Transport Verified
- Environmental Zone
- Truck Toll (for example the ÉcoTaxe of France) - this seems like it's just all highways in France?

**Comment by musculman** [comment]
@nvkelso  thanks for following up on this one.

Let's look first at the questions you mention:

Regarding the `Truck Speed Limit in km/h` property, it can be also used for *visual display* for example when visualizing truck only road shields like [this one from the US](https://en.wikipedia.org/wiki/Speed_limits_in_the_United_States#/media/File:Trucks_speed_55_sign.svg). Check also the example shown in the [Restrictions page of OSM](https://wiki.openstreetmap.org/wiki/Restrictions) under [maxspeed](https://wiki.openstreetmap.org/wiki/Key:maxspeed).

Regarding `environmental zone` property, this one is for *visual display* as well as some cities have areas that can only be accessed by vehicles with certain environmental certification, for example, London, and this is visualized in the map, take a look at [this example from HERE Map Tile API](https://developer.here.com/documentation/map-tile/topics/example-truck-congestion.html).

Regarding `truck toll`, it does not only refer to all highways in France, this one was just an example. This refers in fact to the type of toll applied to the road segment, if any, that is applicable for trucks. For example in France all highways indeed would have a toll, and within the truck attributes, the toll type of ÉcoTaxe, some other countries apply different tolls for trucks as opposed to other vehicules, but in fact, this property is only for *routing* so we can ignore it for now.

Regarding `transport verified`, this one is also for *routing only* so we can skip it for now.

I worry a bit about the implementation and how slow can it be to process it if you want to visualize for example the roads that contain some kind of restriction in red, and on top of that show some road shields, you would need to process all segments and then combine the ones matching certain properties.
But in any case, let's see if I can take your answer and adapt the draft I was preparing, feel free to edit this one.

## Roads (Transportation)
(...)
### Road properties (common optional)
(...)
* `truck_restriction` - a `boolean` value indicating if there is any kind of restriction applied to this segment of the road. Only applicable for `kind` values `major_road`, `minor_road`, `path`.
* `truck_hazmat_permit`- a `boolean` value indicating whether a special permit for hazardous materials is required to legally drive on this road.
* `truck_speed_limit` - maximum speed value for trucks circulating in this road, in Km/h.
* `environmental_zone` - a `boolean` value indicating whether this road belongs to a protected environmental area, such as the [London Loz Emission Zone](https://en.wikipedia.org/wiki/London_low_emission_zone).

### Road Transportation `kind_detail` values for `truck_restriction`

Each `truck_restriction` can have one or more of the following `kind_detail` values.

* `truck_restriction_weight` - maximum weight admitted in the road, in kg
* `truck_restriction_height` - maximum height admitted in the road, in cm
* `truck_restriction_length` - maximum length admitted in the road, in cm
* `truck_restriction_width` - maximum width admitted in the road, in cm
* `truck_restriction_wpa` - Weight per Axle, in kg, see description below
* `truck_restriction_kpra` - King pin to Rear Axle Lenght, in cm, see description below

#### Weight per Axle and King pin to Rear Axle

This property shows the maximum weight admitted in the road depending on the number of axles. In some cases, it might be showing information for one, two or more axles. According to the [Longer Heavier Vehicle description](https://en.wikipedia.org/wiki/Longer_Heavier_Vehicle) the maximum is 6 axles. We should anyway not assume this is a hard limit.

Sample restriction input:
```
weight per axle on the main road network
single axle 10.0t
drive axle 11.5t
tandem drive axle with a distance between the axles of
- 1.20m - 1.29m 16.0t
- 1.30m - 1.80m 18.0t
- >1.80m 20.0t
```

Comments:
* I am not sure we need the `truck_network` and `all_truck_networks` ones. I do not know of what could we put in those properties, but probably I am missing something.
* Each `truck_restriction` can indeed have 1+ restrictions so I used your second proposal.
* The Kingpin to Rear Axle Distances seems to be state dependent information and should be encoded first in OSM or other Data Source to be able to pull it in. Perhaps can be skipped for now.

**Comment by nvkelso** [comment]
I'm worried that we'd be segmenting up geometries that are meant for **visual display** (not database storage) with information that's only useful for truck **routing**.

Discussion:

- Seems like **hazmat** is similar to other types of restriction so I think that should roll under `truck_restriction` and allow that to be text values, including a "multiple" option?
- The OSM-ism for property name is "heavy goods vehicle" so we should use the abbreviation of **hgv** instead of _truck_.

**Yes:**

* `hgv` an optional property indicating general truck heavy goods vehicle (truck) access. Values include: `no`, `designated`, `destination`, `delivery`, `local`, and `agricultural`.
* `hgv_restriction` - an optional property indicating limitations to heavy goods vehicle (truck) access. Values include: `weight`, `height`, `length`, `width`, `wpa`, `kpra`, `hazmat`, and `multiple` if more than one.

Maybe:

* **environmental_zone** - a polygon indicating area of roads inside protected environmental area, such as the [London Loz Emission Zone](https://en.wikipedia.org/wiki/London_low_emission_zone).
    - I don't think we need to set properties on each and every road segment, that's a routing property.
    - Either this goes in landuse layer or roads layer.
* **toll** to indicate boolean that a fee must be paid by general traffic to use a road, road bridge or road tunnel. We don't have this now, but we should for general car use.
    *  **toll_hgv** as a boolean when toll only applies to certain classes of vehicles like trucks.

No: 

- `hgv_speed_limit`: Tilezen purposely doesn't support truck speed limit display for non-truck modes of travel so I don't think we should add this.

I'm still not convinced on these properties:

* `hgv` - for POINTS an optional property indicating limitations to heavy goods vehicle (truck) access. Values include: `weight`, `height`, `length`, `width`, `wpa`, `kpra`, `hazmat`, and `multiple` if more than one. 
    - Maybe these are maximum min_zoom only (eg zoom 16) and not in earlier tiles?
        - If it's a significant restriction either based on the restriction or the road type, show them earlier?
    - These feel more like traffic incident points that just need a title and description text strings...
    - `hgv_weight` - maximum weight admitted in the road, in kg
    - `hgv_height` - maximum height admitted in the road, in cm
    - `hgv_length` - maximum length admitted in the road, in cm
    - `hgv_width` - maximum width admitted in the road, in cm
    - `hgv_wpa` - Weight per Axle, in kg, see description below
    - `hgv_kpra` - King pin to Rear Axle Lenght, in cm, see description below

It'd be nice to see a prototype of what this would actually look like visually and in the tiles.

**Comment by musculman** [comment]
* `hazmat` proposal: makes sense
* using `hgv` instead of  `truck` is a good idea, I was wondering the same when I found the OSM wiki and started to read about it there.
* moving `environmental_zone` to other layer: sure, wherever it makes more sense, after all usually it is stored as a pre-calculated polygon in the tiles, otherwise it is too expensive to do it on the fly on the client.
* proposal for `toll`: makes sense

As for the last part, which seems to be the most difficult as for the others we pretty much agree.
I think that yes, seeing how this looks like in a real map helps.
Because most of these point features are shown usually on zoom levels before 16 in the form of shields that display these special attributes. For example length, height, wpa...
You can see here an [example from HERE Map Tile API](https://developer.here.com/documentation/map-tile/topics/truck-information.html) just for a single tile, but you can use any viewer that can render raster tiles to visualize more zoom levels.

**Comment by nvkelso** [comment]
Looking at the example from the HERE API:

![image](https://user-images.githubusercontent.com/853051/46629414-93301400-caf5-11e8-9902-ca53dd9704db.png)

There is a general icon for what _kind_ of restriction, and then there is text as part of that symbol to provide details on the width or heigh or other value. 

In Tilezen we capture that with **`shield_text`** property with values like `5.5m`. Then those two are combined client side in Tangram or another graphics library.

To update the earlier proposal:

- `hgv_restriction` - kind for POINTs  indicating limitations to heavy goods vehicle (truck) access. Related `kind_detail` values include: `weight`, `height`, `length`, `width`, `wpa`, `kpra`, `hazmat`, and `multiple` if more than one. See also: `hgv_restriction_shield_text`.
- `hgv_restriction_shield_text` - paired with `hgv_restriction` points with values like `5.1m`. Because the units are different per restiction an appreviation should be provided. Values in meters can be specified with one decimal precision (but value of 5.0m should be given as 5m).

> moving environmental_zone to other layer: sure, wherever it makes more sense, after all usually it is stored as a pre-calculated polygon in the tiles, otherwise it is too expensive to do it on the fly on the client.

Let's use **landuse** layer instead of roads, and indicate in this new documentation there's an optional polygon geometry in the landuse layer.

**Comment by musculman** [comment]
I think the `shield_text` part is clear but I still have issues with the general restrictions, and sorry for that if I am repeating myself, just want to make sure it would work as I expect it would.
If you see the tile above, there are two kinds of information:

1.  some parts of the road network are colored red (they contain at least one restriction)
2. there are some shields displayed with information relevant for HGV

As the road network in item 1 can start/finish in any road segment and finalize in any other place, taking or not into account other streets or roads that intersect; it is very difficult to implement it with `point `feature as it is not really a valid description of "show read from point A to B" but rather "show red on the following segments affected by these restrictions".
I just realized that the mechanics should work in a similar way as `traffic_flow` as opposed to `traffic_incidents`, so, geometry controlled vs points. Refer to #1598 
For item 2 we can use for sure `point `feature.

Alternatively, I perhaps misunderstood your proposal :)

**Comment by nvkelso** [comment]
> As the road network in item 1 can start/finish in any road segment and finalize in any other place, taking or not into account other streets or roads that intersect; it is very difficult to implement it with pointfeature as it is not really a valid description of "show read from point A to B" but rather "show red on the following segments affected by these restrictions".

### For item 1

I expect the following new properties on the existing **roads** layer line geometries features:

* `hgv` an optional property indicating general truck heavy goods vehicle (truck) access. Values include: `no`, `designated`, `destination`, `delivery`, `local`, and `agricultural`.
* `hgv_restriction` - an optional property indicating limitations to heavy goods vehicle (truck) access. Values include: `weight`, `height`, `length`, `width`, `wpa`, `kpra`, `hazmat`, and `multiple` if more than one.
* `toll` to indicate boolean that a fee must be paid by general traffic to use a road, road bridge or road tunnel. 
*  `toll_hgv` as a boolean when toll only applies to certain classes of vehicles like heavy good vehicle trucks.

### For item 2

I expect new point geometry features in the existing **roads** layer:

- new `hgv_restriction` **kind** - indicate point location for access limitations for heavy goods vehicles (trucks). 
- new **kind_detail** values for `hgv_restriction` include: `weight`, `height`, `length`, `width`, `wpa`, `kpra`, `hazmat`, `other`, and `multiple` if more than one. See also: `hgv_restriction_shield_text`.
- `hgv_restriction_shield_text` - optional property for `hgv_restriction` kind points with values like `5.1m`. Because the units are different per restriction an abbreviation should be provided. Values in meters can be specified with one decimal precision (but value of 5.0m should be given as 5m).

_Note I added `other` as a new `kind_detail` value._

**Comment by musculman** [comment]
All clear now, thank you for the detailed explanation. I think it is in a pretty good shape already, as far as I can see for a first version. Improvements can come later :)

**Comment by nvkelso** [comment]
https://github.com/tilezen/vector-datasource/pull/1709 covers the documentation, but holding this issue open to cover the actual implementation in the Tilzen default schema.

**Comment by nvkelso** [comment]
A height of `150m` on the US50 in Maryland doesn't hold water (though it crosses it at 9/39.1297/-76.8583):

![image](https://user-images.githubusercontent.com/853051/52192861-feb69380-2800-11e9-8dbc-440fb1965d6b.png)

**Comment by nvkelso** [comment]
```
        z-hgv:
            filter: { hgv: true }
            draw: 
                lines: 
                    visible: true
                    color: blue
                    width: 1.7px
            z-restricted:
                filter:
                    - hgv_restriction: true
                    - hgv: [no]
                draw: 
                    lines: 
                        color: red
                        width: 2.5px
                z-shields:
                    filter: { hgv_restriction_shield_text: true }
                    draw:
                        mapzen_icon_library:
                            # you need to match any custom shield to the vector tile `network` values
                            sprite: |
                                function() {
                                    if( feature.hgv_restriction_shield_text ) {
                                        return ('generic_shield-' + feature.hgv_restriction_shield_text.length + 'char');
                                    }
                                }
                            size: [[7,[75%,75%]],[12,[100%,100%]],[15,[110%,110%]]]
                            priority: 1
                            # Size is specified below, and limited to 5 char in another filter
                            repeat_group: shields
                            repeat_distance: [[8,90px],[10,40px],[13,30px]]
                            placement: midpoint
                            color: red
                            placement_min_length_ratio:
                                - [8, 0.1]
                                - [9, 0.25]
                                - [10, 0.25]
                                - [11, 0.15]
                                - [12, 1]
                                - [13, 1.50]
                                - [14, 2.0]
                            #cull_from_tile: true
                            visible: true
                            text:
                                repeat_distance: 200px
                                anchor: center
                                text_source: hgv_restriction_shield_text
                                font:
                                    family: Montserrat
                                    weight: 500
                                    fill: white
                                    size: [[7, 7px], [13, 9px], [15, 10px]]
                                    stroke: { color: global.text_stroke, width: 0px }
            toll:
                filter: { toll: true }
                draw: 
                    lines: 
                        outline:
                            color: black
                            width: 1.5px
```

![image](https://user-images.githubusercontent.com/853051/52193274-4807e280-2803-11e9-954d-db1f39c8bf27.png)

![image](https://user-images.githubusercontent.com/853051/52193283-5655fe80-2803-11e9-913b-a20b7d8ca92d.png)

![image](https://user-images.githubusercontent.com/853051/52193371-c19fd080-2803-11e9-9e6d-8751df699be7.png)

**Comment by nvkelso** [comment]
The toll information added for this issue is only there for OSM zooms, but Natural Earth also has a toll property. Let's add that for low zooms in the next release.

**Comment by nvkelso** [comment]
Looks good!
